lib/irq.h

```diff
@@ -95,6 +143,158 @@ void metal_irq_enable(unsigned int vector);
  */
 void metal_irq_disable(unsigned int vector);
 
+#define METAL_IRQ_CONTROLLER_DECLARE(_irq_controller, \
+				     _irq_base, _irq_num, \
+				     _arg, \
+				     _irq_set_state, \
+				     _irqs) \
+	struct metal_irq_controller _irq_controller = { \
+		.irq_base = _irq_base, \
+		.irq_num = _irq_num, \
+		.arg = _arg, \
+		.irq_set_state = _irq_set_state, \
+		.irqs = _irqs,\
+	};
+
+/**
+ * @brief	metal_irq_register_controller
+ *
+ * Register IRQ controller
+ *
+ * @param[in] cntr  Interrupt controller to register
+ * @return 0 on success, or negative value for failure.
+ */
+void metal_irq_register_controller(struct metal_irq_controller *cntr);
+
+/**
+ * @brief	metal_irq_allocate
+ *
+ * Allocate IRQ range
+ *
+ * @param[in] irq_base Base of required IRQ range, it can be METAL_IRQ_ANY.
+ *                     If it is METAL_IRQ_ANY, libmetal will allocate the
+ *                     base or IRQ range.
+ * @param[in] irq_num  Number of IRQs required
+ * @return irq number base for success, or negative value for failure.
+ */
+int metal_irq_allocate(int irq_base, int irq_num);
```

Comment:
Do we need a separate function for this?  I would prefer this to be done as part of the metal_irq_register_controller().  This will reduce the number of APIs and I believe metal_irq_register_controller() needs to check that there is no overlap in IRQ ranges anyway.

Reply:
Agree, remove this  function, and let the register to allocate IRQ if required.